# 4. Sets

The final data structure we're going to take a look at in this folder is `Sets`. A set combines some of the features of both `lists` and the `dictionaries`. 

A set is defined as an **unordered**, **mutable** collection of **unique** **items**. A `set` is a data structure where you can store items only once in the structure, without caring about their order.

This description, while highly informal, is rather spot on. Sets in Python are actually analogous to sets in math. For this reason, much of the jargon and functionality that you will see when learning and talking about Python sets is similar to, if not exactly the same as, what applies to mathematical sets.

 [Here](https://en.wikipedia.org/wiki/Set_(mathematics)) is the wikipedia on sets if you want a quick overview of them.

## Objectives

At the end of this notebook you should be able to:

- Understand the difference between sets and lists

- Apply mathematical operations to sets

- Know when to use sets

Let's take a look at how we can construct sets. The first way you could of already guessed by now. The second way not so much.

In [4]:
my_empty_set = set() # Creates an empty set that can be filled later
print(type(my_empty_set))

<class 'set'>


In [5]:
my_set = set([1, 2, 3]) # Creates a set from a list.
my_set

{1, 2, 3}

**Note** that sets only contain unique elements, so duplicates will be removed.

Simply using curly braces is a more concise way to create a set, but it cannot be used to create an empty set as `{}` creates an empty dictionary by defualt.

In [11]:
my_other_set = {1, 2, 2, 3, 3, 3} # Creates a set using curly braces. 
my_other_set

{1, 2, 3}

Sets with the same items in them will evaluate as equal or `True` using boolean logic.

In [8]:
# We can create identical sets:
print(my_set)
print(my_other_set)
print(my_set == my_other_set)

{1, 2, 3}
{1, 2, 3}
True


Above, we see the two ways to make sets:

1. We can use the `set` constructor, which takes an iterable as an arguement
2. We can use the **syntactic sugary** curly braces

**Note**: The curly braces are also used for dictionaries. But, with dictionaries, a colon is used to separate the keywords and values. This is how Python determines whether or not you're declaring a set or a dictionary. 

The only place where Python wouldn't know is when we declare an empty structure. When this happens, Python assumes the empty curly braces `{}` always mean an empty dictionary. 

If we take a look at the methods that are available on sets we see:
```python
set.add                          set.intersection                 set.remove
set.clear                        set.intersection_update          set.symmetric_difference
set.copy                         set.isdisjoint                   set.symmetric_difference_update
set.difference                   set.issubset                     set.union
set.difference_update            set.issuperset                   set.update
set.discard                      set.pop  
```

As mentioned earlier, many of the methods available to python sets are similar to, if not the same as, those available to mathematical sets. Naturally, we see ways to compute set operations (`intersection()`, `union()`, etc.) and alter the set (`add()`, `update()`, `pop()` and `remove()`). 

Let's take a look at some of these methods in action and try to figure out what they do.

In [12]:
my_set, my_other_set = {1, 2, 3}, {5, 6, 7}

In [13]:
my_set.union(my_other_set)

{1, 2, 3, 5, 6, 7}

In [14]:
my_set.add(4)
my_set

{1, 2, 3, 4}

In [15]:
my_set.update(my_other_set)
my_set

{1, 2, 3, 4, 5, 6, 7}

In [16]:
my_set.remove(5)
my_set

{1, 2, 3, 4, 6, 7}

In [17]:
my_set.intersection(my_other_set)

{6, 7}

All of these methods should look fairly intuitive. 

- The `update()` method is like an `add()` en masse
- The `union()` method is like adding two sets together, but since there are only unique elements in a set, it removes duplicates
- The `intersection()` method returns those elements that the sets have in common
- The `remove()` method removes elements....

These are some of the most common set operations you will ever use. If you'd like to take a look at the documentation for all of them, check it out [Here](https://docs.python.org/2/library/stdtypes.html#set).

#### Why Do We Need Sets?

We have already covered many different data types in python, they should be enough.... so **when** would I use a set? 

The most obvious answer is for times when you need to perform set operations. 

For example: 

1. Checking what **elements** **two lists** have in common by taking a set of them both and finding the intersection of those sets. 
2. To find the **unique** items in an **iterable**. 

There is also another example where we'll want to use sets that is not so apparent.

3. If we want to check if an item is in a list, it requires us to check every item in that list. This can be computationally expensive especally if the list is very big. Instead, we use a set! 

The reason why we use a set lies in the fact that sets in Python are built very similarly to dictionaries. There's an underlying hash table that allows elements to be stored and queried for membership in the set (**Note**, *this means that the elements of a set have to be immutable*). 

This operation happens much faster with sets than with lists [Here](https://wiki.python.org/moin/TimeComplexity) is some coverage on how quickly some Python methods run, just to show you how useful sets really are. 

Let's take a look at this in action, and simultaneously learn about how to time things in IPython.

In [18]:
my_list = list(range(1000000))

In [19]:
my_set = set(my_list)

In [22]:
timeit 100000 in my_list

839 μs ± 408 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [21]:
timeit 100000 in my_set

25.2 ns ± 0.179 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


Here, we used the `timeit` function that's built into Jupyter. To use it is very simple. Just like in the above cells, call `timeit` at the start of your line of code. 

We can see that the `list` version of checking memebership in a collection took approx. 839 μs ± 408 ns per loop, and for the `set` version approx. 25.2 ns ± 0.179 ns per loop. These numbers will vary from computer to computer but the scale should remain consistent.

Just to put these scales into perspective:

- 839 μs (microseconds) = 0.000839 s
- 25.2 ns (nanoseconds) = 0.0000000252 s 

Lets do some maths....

In [27]:
print(f"A set is {round(0.000839 / 0.0000000252, 1)} times faster than the list")

A set is 33293.7 times faster than the list


The `list` version of checking membership in a collection took ~30000 times longer than the `set` version. That number is proportional to the size of the collection and will only increase if the collection increases. 

So, in the future, if you are ever working with vast amounts of data, consider using sets. They can save you a lot of time.  

## Check your Understanding:

**Part 1**

1. Make a set called `first_set` with the values 1-10 and another with the values 5-15 called `second_set`
2. Add the value 11 to `first_set`
3. Add the string `'hello'` to `second_set`
4. Using one of the methods discussed above, find what elements `first_set` and `second_set` have in common
5. In one line of code, add all the elements of `second_set` to `first_set`

In [ ]:
# Answer Here

**Part 2**

Given the lists:

`first_list = ['hello', 'there', 'things', 'stuff', 'other', 'soda', 'chicken wings', 'things', 'soda']` 

`second_list = ['turkey sandwich', 'guacamole', 'chicken wings', 'OJ', 'soda']`

6. Find the unique elements in `first_list`
7. using one line of code, find the common elements in the two lists
8. Write a single line of code that outputs `True` or `False` depending on if the string `'pizza'` is in both lists.

In [ ]:
# Answer Here